onnx/defs/nn/defs.cc

```diff
@@ -2832,4 +2833,392 @@ ONNX_OPERATOR_SET_SCHEMA(
+              .Const1D("Zero1D", static_cast<int64_t>(0))
+              .Add("CalculatedScale = Div(One1D, SqrtHeadSize)")
+              .Const("ScaleF", ToTensor<float>(scale))
+              .Add(scale_attr != nullptr ? "ScaleFactor = Identity(ScaleF)" : "ScaleFactor = Identity(CalculatedScale)")
+              .Add("ScaleFactorF = Cast (ScaleFactor)", "to", T);
+
+          // Update key and value caches for past and present states
+          if (ctx.hasOutput(1)) {
+            if (ctx.hasInput(4)) {
+              builder.Add("present_key = Concat <axis = 2> (past_key, KReshaped)");
+            }
+          } else {
+            builder.Add("present_key = Identity (KReshaped)");
+          }
+
+          if (ctx.hasOutput(2)) {
+            if (ctx.hasInput(5)) {
+              builder.Add("present_value = Concat <axis = 2> (past_value, VReshaped)");
+            }
+          } else {
+            builder.Add("present_value = Identity (VReshaped)");
+          }
+
+          // Create a attn_bias filled with zeros of shape (q_sequence_length, kv_sequence_length)
+          builder.Add("NewKVSeqLen =  Shape <start = -2, end = -1> (present_key)")
+              .Add("AttnBiasShape = Concat <axis = -1> (QSeqLen, NewKVSeqLen)")
+              .Add("AttnBiasZeros = ConstantOfShape(AttnBiasShape)");
+
+          // If attn_mask is provided
+          if (ctx.hasInput(3)) {
+            auto* up = ctx.getInputType(3);
+            if ((up == nullptr) || (!up->has_tensor_type()))
+              return false;
+            int64_t U = up->tensor_type().elem_type();
+            builder
+                .Add(
+                    U == ONNX_NAMESPACE::TensorProto_DataType_BOOL
+                        ? "MaskedAttnBias = Where(attn_mask, AttnBiasZeros, FloatInf)"
+                        : "MaskedAttnBias = Add(attn_mask, AttnBiasZeros)")
+                .Add("AttnBiasT = Cast (MaskedAttnBias)", "to", T);
+          } else {
+            // If is_causal set to true, the attention masking is a lower triangular matrix when the mask
+            // is a square matrix. The attention masking has the form of the upper left causal bias due to
+            // the alignment when the mask is a non-square matrix.
+            // An error is thrown if both attn_mask and is_causal are set.
+            auto* is_causal_attr = ctx.getAttribute("is_causal");
+            int64_t is_causal = (is_causal_attr != nullptr) ? is_causal_attr->i() : 0;
+            float neg_inf = -std::numeric_limits<float>::infinity();
+            builder.Add("TempMask = ConstantOfShape(AttnBiasShape)", "value", mktensor(1))
+                .Add("TempMaskTri = Trilu <upper = 0> (TempMask, Zero1D)")
+                .Const1D("FloatInf", neg_inf)
+                .Add("CasualAttnBias = Where(TempMaskTri, AttnBiasZeros, FloatInf)")
+                .Const("IsCasual", is_causal)
+                .Add("AttnBias = Where(IsCausal, CasualAttnBias, AttnBiasZeros)")
```

Comment:
I think this can be simplified if we generate different bodies for the is_causal and not is_causal cases. No need to encode the condition in the generated body.